tests/draft4/ref.json

```diff
@@ -296,5 +296,110 @@
                 "valid": false
             }
         ]
+    },
+    {
+        "description": "Location-independent identifier",
+        "schema": {
+            "$ref": "#foo",
+            "definitions": {
+                "A": {
+                    "id": "#foo",
+                    "type": "integer"
+                }
+            }
+        },
+        "tests": [
+            {
+                "data": 1,
+                "description": "match",
+                "valid": true
+            },
+            {
+                "data": "a",
+                "description": "mismatch",
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "Canonical location-independent identifier",
+        "schema": {
+            "$ref": "http://localhost:1234/bar#foo",
+            "definitions": {
+                "A": {
+                    "id": "http://localhost:1234/bar#foo",
+                    "type": "integer"
+                }
+            }
+        },
+        "tests": [
+            {
+                "data": 1,
+                "description": "match",
+                "valid": true
+            },
+            {
+                "data": "a",
+                "description": "mismatch",
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "Location-independent identifier with base URI change in subschema",
+        "schema": {
+            "id": "http://localhost:1234/root",
+            "$ref": "http://localhost:1234/nested.json#foo",
```

Comment:
Does having `id` adjacent to `$ref` work?  I'd really rather wrap the `$ref` in an `allOf`.  I know that within the spec we allow adjacent `definitions`, and I suppose that argument could be extended to `id`.  But really it's a gray area/bug in the spec and implementations are not consistent about whether it's supported.